extensions/stats/config.h

```diff
@@ -0,0 +1,384 @@
+};
+
+using valueExtractorFn = uint64_t (*)(const Common::RequestInfo& request_info);
+
+class SimpleStat {
+ public:
+  SimpleStat(uint32_t metric_id, valueExtractorFn value)
+      : metric_id_(metric_id), valueFn_(value){};
+
+  inline void record(const Common::RequestInfo& request_info) {
+    recordMetric(metric_id_, valueFn_(request_info));
+  };
+
+ private:
+  uint32_t metric_id_;
+  valueExtractorFn valueFn_;
+};
+
+using SimpleStatSharedPtr = std::shared_ptr<SimpleStat>;
+using MetricSharedPtr = std::shared_ptr<Metric>;
+
+#define SYMIFEMPTY(ex, sym) (ex).empty() ? (sym) : (ex)
+
+using mapperFn = std::function<std::string(
+    bool outbound, const common::NodeInfo& source, const common::NodeInfo& dest,
+    const Common::RequestInfo& requestInfo)>;
+
+struct Mapping {
+ public:
+  Mapping(std::string name, mapperFn mapper) : name_(name), mapper_(mapper){};
+
+  Mapping() = delete;
+
+  std::string name_;
+  mapperFn mapper_;
+};
+std::vector<Mapping> getStandardMappings();
+
+class Mappings {
+ public:
+  Mappings(std::vector<Mapping> mappings) : mappings_(mappings){};
+
+  Mappings() : Mappings(getStandardMappings()) {}
+
+  // metricTags converts mappings into ordered tags
+  std::vector<MetricTag> metricTags() {
+    std::vector<MetricTag> ret;
+
+    ret.reserve(mappings_.size());
+    for (const auto& mapping : mappings_) {
+      ret.push_back({mapping.name_, MetricTag::TagType::String});
+    }
+    return ret;
+  }
+
+  std::vector<std::string> eval(bool outbound, const common::NodeInfo& source,
+                                const common::NodeInfo& dest,
+                                const Common::RequestInfo& requestInfo) {
+    std::vector<std::string> vals;
+
+    vals.reserve(mappings_.size());
+    for (const auto& mapping : mappings_) {
+      vals.push_back(mapping.mapper_(outbound, source, dest, requestInfo));
+    }
+    return vals;
+  }
+
+ private:
+  std::vector<Mapping> mappings_;
+};
+
+#define MAPPING_SYM(key, expr, sym)                                      \
+  {                                                                      \
+    (key),                                                               \
+        [](bool ABSL_ATTRIBUTE_UNUSED outbound,                          \
+           const common::NodeInfo& ABSL_ATTRIBUTE_UNUSED source,         \
+           const common::NodeInfo& ABSL_ATTRIBUTE_UNUSED dest,           \
+           const Common::RequestInfo& ABSL_ATTRIBUTE_UNUSED requestInfo) \
+            -> std::string {                                             \
+          auto source_labels = source.labels();                          \
+          auto dest_labels = dest.labels();                              \
+          auto val = (expr);                                             \
+          logDebug(absl::StrCat((key), "=", ToString(val)));             \
+          return (SYMIFEMPTY(ToString(val), (sym)));                     \
```

Comment:
kind of complex here but I think ToString execute twice

Reply:
It would be simpler to define a bunch of static functions instead of building lambdas. It is a lot easier to debug functions with names.

Reply:
Also, the arguments should be collected into a context object here I think, and captured by reference.